ops/lib/__init__.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+import re
+
+from pkgutil import get_importer
+from importlib.util import module_from_spec
+from functools import total_ordering
+
+
+_libraries = {}
+
+_libline_re = re.compile(r'''^LIB([A-Z]+)\s+=\s+([0-9]+|['"][a-zA-Z0-9_.-@]+['"])\s*$''')
+_libname_re = re.compile(r'''^[a-z][a-z0-9]+$''')
+
+# Not perfect, but should do for now.
+_libauthor_re = re.compile(r'''^[A-Za-z0-9_+.-]+@[a-z0-9_-]+(?:\.[a-z0-9_-]+)*\.[a-z]{2,3}$''')
+
+
+def use(name, author, api):
+    if not _libname_re.match(name):
+        raise ValueError("invalid library name: {} (chars and digits only)".format(name))
+    if not _libauthor_re.match(author):
+        raise ValueError("invalid library author email: {}".format(author))
+    if not isinstance(api, int):
+        raise ValueError("invalid library API: {} (must be an int)".format(api))
```

Comment:
Any chance we can support strings instead so that we can use SemVer-style <major>.<minor>.<patch> versions?

Reply:
This is inspired by semantic versioning, but goes a step further by being more clear about the intention of the numbers so that people will have a harder time breaking what's encouraged by SemVer. API version is the Major.

Reply:
Thanks for the clarification. I'm guessing the Major-version-only adoption of SemVer here is inspired by the way REST/HTTP-based APIs out there (e.g. facebook, twitter) do it. It works well for those single-instance APIs but I wonder if this might cause problems in this context where we are dealing with libraries that have copies everywhere (in production, in dev 1's workstation, in dev 2's workstation, in some random user trying out the charm). In that case, relying on just the Major version may be a problem because dev 1 might have v1.3 which has the features that the charm needs but dev 2's might have v1.0 which doesn't have that same version. But because we only care about the major version, this tool won't report any problems. I don't know what the chances of that are happening and how expensive that is to debug and resolve but thought I'd throw that scenario out there for your consideration.

Reply:
It's not relying just on major version, and such concerns are addressed. I suggest actually reviewing the code, or talking to the team to get a bit more insight.

Reply:
Thanks for the quick responses @niemeyer! Really appreciate it. I'm about halfway through the code and jumping between that and the Operator Framework Libraries doc that describes its uses and design principles. I'm also going to be picking the team's brain next week about it. In the meantime, have a great weekend!